cdap-messaging-ext-spanner/src/main/java/io/cdap/cdap/messaging/spanner/SpannerMessagingService.java

```diff
@@ -0,0 +1,367 @@
+import io.cdap.cdap.messaging.spi.TopicMetadata;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.id.TopicId;
+import io.cdap.cdap.security.spi.authorization.UnauthorizedException;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutionException;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SpannerMessagingService implements MessagingService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SpannerMessagingService.class);
+  public static final String PAYLOAD_FIELD = "payload";
+  public static final String PUBLISH_TS_FIELD = "publish_ts";
+  public static final String PAYLOAD_SEQUENCE_ID = "payload_sequence_id";
+  public static final String SEQUENCE_ID_FIELD = "sequence_id";
+  public static final String TOPIC_METADATA_TABLE = "topic_metadata";
+  public static final String TOPIC_ID_FIELD = "topic_id";
+  public static final String PROPERTIES_FIELD = "properties";
+  public static final String NAMESPACE_FIELD = "namespace";
+  public static final String TOPIC_TABLE_PREFIX = "messaging";
+
+  private DatabaseClient client;
+
+  private DatabaseAdminClient adminClient;
+
+  private String instanceId;
+
+  private String databaseId;
+
+  private int publishBatchSize;
+
+  private int publishBatchTimeoutMillis;
+
+  private int publishDelayMillis;
+
+  private final ConcurrentLinkedQueue<StoreRequest> batch = new ConcurrentLinkedQueue<>();
+
+  @Override
+  public void initialize(MessagingServiceContext context) throws IOException {
+    Map<String, String> cConf = context.getProperties();
+    this.databaseId = SpannerUtil.getDatabaseID(cConf);
+    this.instanceId = SpannerUtil.getInstanceID(cConf);
+    String projectID = SpannerUtil.getProjectID(cConf);
+    Credentials credentials = SpannerUtil.getCredentials(cConf);
+
+    this.publishBatchSize = Integer.parseInt(cConf.get(SpannerUtil.PUBLISH_BATCH_SIZE));
+    this.publishBatchTimeoutMillis = Integer.parseInt(
+        cConf.get(SpannerUtil.PUBLISH_BATCH_TIMEOUT_MILLIS));
+    this.publishDelayMillis = Integer.parseInt(cConf.get(SpannerUtil.PUBLISH_BATCH_POLL_MILLIS));
+
+    Spanner spanner = SpannerUtil.getSpannerService(projectID, credentials);
+    this.client = SpannerUtil.getSpannerDbClient(projectID, instanceId, databaseId, spanner);
+    this.adminClient = SpannerUtil.getSpannerDbAdminClient(spanner);
+    LOG.info("Spanner messaging service started.");
+
+    createTopics(context.getSystemTopics());
+  }
+
+  @Override
+  public String getName() {
+    return this.getClass().getSimpleName();
+  }
+
+  @Override
+  public void createTopic(TopicMetadata topicMetadata)
+      throws TopicAlreadyExistsException, IOException, UnauthorizedException {
+    List<String> ddlStatements = new ArrayList<>();
```

Comment:
You don't need additional implementation here. Simply call createTopic(ImmutableList.of(topicMetadata)).